From: Maximilian Weigand <mweigand@mweigand.net>
Date: Wed, 7 Dec 2022 09:45:08 +0100
Subject: [rockchip-ebc] Try to improve latency of the refresh thread

When measuring individual frame-processing times it became evident that
some frames took way longer than others (usually frame processing takes
about 15 ms, including waking up from the completion, but 30 ms
frame-processing time was sometimes observed). This is mainly caused by
the use of the spinlock, which encompasses large sections of code both
in the partial refresh function and the atomic_update function (I think
it is generally suggested to use mutexes in those cases - to be
investigated at another time).

The mitigation strategy implemented in this patch is as follows: In the
partial refresh thread, use spin_trylock to acquire the lock. If this
fails, no new clips will be processed in this frame, but frame execution
will be more consistent. To give the partial_refresh thread a better
chance to get the lock without delay, add some delays to the
atomic_update function AFTER releasing the lock there. This prevents
overly eager applications to swamp the update queue. Note that the
choice of delay time is the result of not very much testing. Too small
values will lead to not enough clips being processed, while too large
values will make everything slower.

My personal observation is that drawing in xournalpp gets somewhat
smoother due to reduced variations in drawing updates. The downside is
that this "optimization" leads to somewhat increased artifacting.

Lets see if we need to revert this patch....
---
 drivers/gpu/drm/rockchip/rockchip_ebc.c | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_ebc.c b/drivers/gpu/drm/rockchip/rockchip_ebc.c
index 26f9196..bad06fc 100644
--- a/drivers/gpu/drm/rockchip/rockchip_ebc.c
+++ b/drivers/gpu/drm/rockchip/rockchip_ebc.c
@@ -17,6 +17,7 @@
 #include <linux/dma-mapping.h>
 #include <linux/uaccess.h>
 #include <linux/firmware.h>
+#include <linux/delay.h>
 
 #include <drm/drm_atomic.h>
 #include <drm/drm_atomic_helper.h>
@@ -877,13 +878,17 @@ static void rockchip_ebc_partial_refresh(struct rockchip_ebc *ebc,
 		bool sync_next = false;
 		bool sync_prev = false;
 		int split_counter = 0;
+		int gotlock;
 
 		// now the CPU is allowed to change the phase buffer
 		dma_sync_single_for_cpu(dev, phase_handle, ctx->phase_size, DMA_TO_DEVICE);
 
 		/* Move the queued damage areas to the local list. */
-		spin_lock(&ctx->queue_lock);
-		list_splice_tail_init(&ctx->queue, &areas);
+		/* spin_lock(&ctx->queue_lock); */
+		/* list_splice_tail_init(&ctx->queue, &areas); */
+		gotlock = spin_trylock(&ctx->queue_lock);
+        if (gotlock)
+			list_splice_tail_init(&ctx->queue, &areas);
 		/* spin_unlock(&ctx->queue_lock); */
 
 		list_for_each_entry_safe(area, next_area, &areas, list) {
@@ -975,7 +980,9 @@ static void rockchip_ebc_partial_refresh(struct rockchip_ebc *ebc,
 		dma_sync_single_for_device(dev, phase_handle,
 					   ctx->phase_size, DMA_TO_DEVICE);
 
-		spin_unlock(&ctx->queue_lock);
+		if (gotlock)
+			spin_unlock(&ctx->queue_lock);
+		/* spin_unlock(&ctx->queue_lock); */
 
 		/* if (frame) { */
 		/* 	if (!wait_for_completion_timeout(&ebc->display_end, */
@@ -1863,12 +1870,18 @@ static void rockchip_ebc_plane_atomic_update(struct drm_plane *plane,
 
 	if (list_empty(&ebc_plane_state->areas)){
 		spin_unlock(&ctx->queue_lock);
+		// the idea here: give the refresh thread time to acquire the lock
+		// before new clips arrive
+		usleep_range(5000, 5500);
 		return;
 	}
 
 	/* spin_lock(&ctx->queue_lock); */
 	list_splice_tail_init(&ebc_plane_state->areas, &ctx->queue);
 	spin_unlock(&ctx->queue_lock);
+	// the idea here: give the refresh thread time to acquire the lock
+	// before new clips arrive
+	usleep_range(5000, 5100);
 
 	wake_up_process(ebc->refresh_thread);
 }
